### PR TITLE
ConfigurationFactory

### PR DESCRIPTION
**Pull request by scottbommarito**
Created `IConfigurationFactory` and its implementation, `ConfigurationFactory`, as the major source for configuration in our projects. It's purpose is to inflate a class with configuration based on a `IConfigurationProvider` that is provided to it.

Using `IConfigurationFactory`, configuration can now be constructed at a single point and thereafter passed around statically. The user of the class has control over when configuration is generated, thus allowing any flow (for example, a single request) to maintain a single configuration throughout. Additionally, all configuration can be accessed statically, so there is no longer any issues with contexts that cannot be asynchronous.

Also, removed the synchronous methods from `IConfigurationProvider` since they were only used for returning cached values. All accesses to actual secret values must be asynchronous. `SecretConfigurationProvider` was removed as well because its uses can be entirely replaced by `ConfigurationFactory`.

### CODE REVIEW

File: src/NuGet.Services.Configuration/Configuration.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+
+namespace NuGet.Services.Configuration
+{
+    /// <summary>
+    /// Base class to extend when using an <see cref="IConfigurationFactory"/>.
+    /// </summary>
+    public class Configuration
```
Comment (skofman1) [inline src/NuGet.Services.Configuration/Configuration.cs]: >Configuration [](start = 17, length = 13)

should be an abstract class

File: src/NuGet.Services.Configuration/IConfigurationProvider.cs [hunk 2]
```diff
@@ -8,51 +8,29 @@
 namespace NuGet.Services.Configuration
 {
     /// <summary>
-    /// Asynchronously provides configuration or command line arguments.
+    /// Asynchronously provides configuration.
+    /// Used when all possible configuration cannot be cached, for example RoleEnvironment.
     /// </summary>
     public interface IConfigurationProvider
```
Comment (skofman1) [inline src/NuGet.Services.Configuration/IConfigurationProvider.cs]: >IConfigurationProvider [](start = 21, length = 22)

I would recommend doing a much simpler interface: one method:

bool TryGet(string key, out string value)

Comment (skofman1) [inline src/NuGet.Services.Configuration/IConfigurationProvider.cs]: >interface [](start = 11, length = 9)

I would recommend doing a much simpler interface: one method:

bool TryGet(string key, out string value)

Since IConfigurationProvider is a wrapper around a RoleEnvironment/json file/environment variable, this is all you need.

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+namespace NuGet.Services.Configuration
+{
+    /// <summary>
+    /// <see cref="IConfigurationFactory"/> implementation that uses an <see cref="IConfigurationProvider"/> to inject configuration.
+    /// </summary>
+    public class ConfigurationFactory : IConfigurationFactory
+    {
+        private readonly IConfigurationProvider _configProvider;
+
+        public ConfigurationFactory(IConfigurationProvider configProvider)
```
Comment (skofman1) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: >configProvider [](start = 59, length = 14)

check for null

**Comment by skofman1** [comment]
> using System;

file headers everywhere

---
Refers to: src/NuGet.Services.Configuration/ConfigurationFactory.cs:1 in 9be2b0f. [](commit_id = 9be2b0f1403a719a971380ef58b351f4d773e8cc, deletion_comment = False)

File: src/NuGet.Services.Configuration/IConfigurationProvider.cs (hunk 2 above)
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/IConfigurationProvider.cs]: `IConfigurationProvider` is a wrapper around a RoleEnvironment/json file/environment variable that prepares the configuration for consumption, which includes asynchronously injecting secrets. Therefore the accessors must be async. `TryGet` does not work with async, since `out` parameters are not possible.

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            }
+            else
+            {
+                var defaultValueAttribute = property.Attributes.OfType<DefaultValueAttribute>().FirstOrDefault();
+                if (defaultValueAttribute != null)
+                {
+                    try
+                    {
+                        // Use the default value specified by the DefaultValueAttribute if it can be converted into the type of the property.
+                        var defaultValue = (TP) (defaultValueAttribute.Value.GetType() == property.PropertyType ? defaultValueAttribute.Value : property.Converter.ConvertFrom(defaultValueAttribute.Value));
```
Comment (skofman1) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: >ty.Pr [](start = 96, length = 5)

break line

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            if (property.Attributes.OfType<RequiredAttribute>().Any())
+            {
+                // If the property is required, use GetOrThrowAsync to access configuration.
+                // It will throw if the configuration is not found or invalid.
+                value = await _configProvider.GetOrThrowAsync<TP>(settingName);
+            }
+            else
+            {
+                var defaultValueAttribute = property.Attributes.OfType<DefaultValueAttribute>().FirstOrDefault();
+                if (defaultValueAttribute != null)
```
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: It grabs the default value and then passes it into the `IConfigurationProvider`. The `IConfigurationProvider` returns the default if it cannot find the value in configuration.

`value = await _configProvider.GetOrDefaultAsync(settingName, defaultValue);`

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+                configKey = configNameProperty.Key;
+            }
+
+            // Add the prefix specified by the ConfigurationKeyAttribute to the configuration key if it exists.
+            var configNamePrefixProperty =
+                property.Attributes.OfType<ConfigurationKeyPrefixAttribute>().FirstOrDefault();
+            if (configNamePrefixProperty != null)
+            {
+                configKey = configNamePrefixProperty.Prefix + configKey;
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: Shouldn't you check for null/empty here like [above](https://github.com/NuGet/ServerCommon/pull/12/files#diff-1f2ec043823e5381e70758cea0bfdbe0R55)?

File: src/NuGet.Services.Configuration/ConfigurationKeyAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+{
+    /// <summary>
+    /// Used to specify a configuration key in a <see cref="Configuration"/> subclass while using a <see cref="ConfigurationFactory"/>.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Property)]
+    public class ConfigurationKeyAttribute : Attribute
+    {
+        public ConfigurationKeyAttribute(string key)
+        {
+            Key = key;
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationKeyAttribute.cs]: Throw if `key` is null or empty.

File: src/NuGet.Services.Configuration/ConfigurationKeyPrefixAttribute.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace NuGet.Services.Configuration
+{
+    /// <summary>
+    /// Used to specify a prefix to add to the configuration key in a <see cref="Configuration"/> subclass while using a <see cref="ConfigurationFactory"/>.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Property)]
+    public class ConfigurationKeyPrefixAttribute : Attribute
+    {
+        public ConfigurationKeyPrefixAttribute(string prefix)
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationKeyPrefixAttribute.cs]: Throw if `prefix` is null or empty.

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                if (string.IsNullOrEmpty(configNameProperty.Key))
+                {
+                    throw new ArgumentNullException(nameof(configNameProperty.Key),
+                        $"Configuration key specified by {nameof(ConfigurationKeyAttribute)} is null or empty!");
+                }
+
+                configKey = configNameProperty.Key;
+            }
+
+            // Add the prefix specified by the ConfigurationKeyAttribute to the configuration key if it exists.
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: `ConfigurationKeyPrefixAttribute`

File: tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs [hunk 10]
```diff
@@ -0,0 +1,184 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (scottbommarito) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs]: This code is identical to what used to be in `SecretConfigurationProviderFacts` but using a `TestConfigurationProvider` instead of the now defunct `SecretConfigurationProvider`.

File: src/NuGet.Services.Configuration/Configuration.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        public Configuration()
+        {
+            CreatedTime = DateTime.UtcNow;
+        }
+
+        public DateTime CreatedTime { get; }
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/Configuration.cs]: Use `DateTimeOffset`.

File: src/NuGet.Services.Configuration/Configuration.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+namespace NuGet.Services.Configuration
+{
+    /// <summary>
+    /// Base class to extend when using an <see cref="IConfigurationFactory"/>.
+    /// </summary>
+    public class Configuration
+    {
+        public Configuration()
+        {
+            CreatedTime = DateTime.UtcNow;
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/Configuration.cs]: Use `DateTimeOffset.UtcNow`.

File: src/NuGet.Services.Configuration/Configuration.cs [hunk 13]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/Configuration.cs]: Add copyright header to all files.

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+        /// <summary>
+        /// Injects a property of <param name="instance">instance</param> specified by <param name="property">a <see cref="PropertyDescriptor"/></param> with configuration.
+        /// </summary>
+        /// <typeparam name="T">Type of <param name="instance">the instance</param>.</typeparam>
+        /// <typeparam name="TP">Type of <param name="property">the property</param>.</typeparam>
+        /// <param name="instance">Instance to inject configuration into a property of.</param>
+        /// <param name="property"><see cref="PropertyDescriptor"/> that describes the property to inject the configuration into.</param>
+        /// <returns>A task that completes when the property has been injected into.</returns>
+        public async Task InjectPropertyWithConfiguration<T, TP>(T instance, PropertyDescriptor property)
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: Throw for `null` arguments.

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+                        var defaultValue =
+                            (TP)
+                            (defaultValueAttribute.Value.GetType() == property.PropertyType
+                                ? defaultValueAttribute.Value
+                                : property.Converter.ConvertFrom(defaultValueAttribute.Value));
+                        value = await _configProvider.GetOrDefaultAsync(configKey, defaultValue);
+                    }
+                    catch (Exception)
+                    {
+                        throw new ArgumentException($"Default value for {configKey} specified by DefaultValueAttribute is malformed ({defaultValueAttribute.Value ?? "null"})!");
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: `{nameof(DefaultValueAttribute)}` like [here](https://github.com/NuGet/ServerCommon/pull/12/files#diff-1f2ec043823e5381e70758cea0bfdbe0R58);

File: src/NuGet.Services.Configuration/ConfigurationProvider.cs [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        /// Gets a value from a given key.
+        /// </summary>
+        /// <param name="key">The key associated with the desired value.</param>
+        /// <returns>The value associated with the given key.</returns>
+        /// <exception cref="KeyNotFoundException">Thrown when the key is not found in the list of keys.</exception>
+        /// <exception cref="ArgumentNullException">Thrown when the value associated with the given key is null or empty.</exception>
+        protected abstract Task<string> Get(string key);
+
+        public async Task<T> GetOrThrowAsync<T>(string key)
+        {
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationProvider.cs]: Throw for `null` or empty `key`.

File: src/NuGet.Services.Configuration/ConfigurationProvider.cs [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+            return ConfigurationUtility.ConvertFromString<T>(await Get(key));
+        }
+
+        public async Task<T> GetOrDefaultAsync<T>(string key, T defaultValue = default(T))
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationProvider.cs]: Throw for `null` or empty `key`.

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+    /// <summary>
+    /// <see cref="IConfigurationFactory"/> implementation that uses an <see cref="IConfigurationProvider"/> to inject configuration.
+    /// </summary>
+    public class ConfigurationFactory : IConfigurationFactory
+    {
+        private readonly IConfigurationProvider _configProvider;
+
+        public ConfigurationFactory(IConfigurationProvider configProvider)
+        {
+            _configProvider = configProvider;
```
Comment (ryuyu) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: Is there any plan to allow multiple configurationProviders to be managed by one factory? Or is the idea that each provider should make its own factory?

Comment (scottbommarito) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: My idea was that if we want to have multiple `IConfigurationProvider`s for one factory, we should create an `AggregateConfigurationProvider` that aggregates a number of `IConfigurationProvider`s and give it to the factory. I haven't found a use for that yet so I didn't write it yet, but it would be fairly easy to do.

Comment (ryuyu) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: So make a new provider that is an aggregation of the providers you need? i guess this makes sense. Sounds fine to me.

File: tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs [hunk 19]
```diff
@@ -0,0 +1,454 @@
...
+            }
+        };
+
+        /// <summary>
+        /// Returns true if the object is convertible to type.
+        /// </summary>
+        /// <param name="value">The object to test.</param>
+        /// <param name="type">Type to test.</param>
+        /// <returns>True if the <param name="value">object</param> is convertible to <param name="type">type</param>, false otherwise.</returns>
+        private static bool IsValueWrongType(object value, Type type)
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs]: Check for valid types instead of invalid types, since the former is a much smaller set than the latter.  The logic will need to be inverted, but I think it'll be simpler and avoid an odd conclusion (`IsValueWrongType(...)` returned `false` so it's a "right" type.).

        private static bool CanConvertTo(object value, Type type)
        {
            if (value == null)
            {
                return !type.IsValueType || Nullable.GetUnderlyingType(type) != null;
            }

            return TypeDescriptor.GetConverter(type).CanConvertFrom(value.GetType());
        }

This will avoid exceptions from failed `ConvertFrom(...)` calls.  If `type` were null, `TypeDescriptor.GetConverter(...)` would throw an `ArgumentNullException`; however, you don't have any such test cases.  Explicitly checking for a `null` `type` would be trivial though.

File: tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs [hunk 20]
```diff
@@ -0,0 +1,454 @@
...
+            {
+                var configTuple = configPair.Value;
+                
+                // True if the expected value is null or if it is an empty string.
+                var expectedValueIsMissing = 
+                    configTuple.ExpectedValue == null ||
+                    (configTuple.ExpectedValue is string && string.IsNullOrEmpty((string) configTuple.ExpectedValue));
+                
+                var expectedIsInvalid = IsValueWrongType(configTuple.ExpectedValue, configTuple.Type);
+                var defaultIsInvalid = IsValueWrongType(configTuple.DefaultValue, configTuple.Type);
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs]: var isExpectedValueValid = IsValueTypeValid(configTuple.ExpectedValue, configTuple.Type);
    var isDefaultValueValid = IsValueTypeValid(configTuple.DefaultValue, configTuple.Type);

File: tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs [hunk 21]
```diff
@@ -0,0 +1,454 @@
...
+        /// Dynamically adds a property to a <see cref="TypeBuilder"/>.
+        /// </summary>
+        /// <param name="typeBuilder">The <see cref="TypeBuilder"/> to add the property to.</param>
+        /// <param name="name">The name of the property.</param>
+        /// <param name="configTuple">Specifies how to construct the property.</param>
+        private static void AddProperty(TypeBuilder typeBuilder, string name, ConfigurationTuple configTuple)
+        {
+            // Create the property attribute.
+            var propertyAttributes = configTuple.DefaultValue != null ? PropertyAttributes.HasDefault : PropertyAttributes.None;
+            var propertyBuilder = typeBuilder.DefineProperty(name, propertyAttributes, configTuple.Type, null);
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs]: `parameterTypes: null`

File: tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs [hunk 22]
```diff
@@ -0,0 +1,454 @@
...
+                        typeof(ConfigurationKeyPrefixAttribute).GetConstructor(new[] { typeof(string) }),
+                        new object[] { configTuple.ConfigKeyPrefix }));
+            }
+
+            // Create the field that the property will get and set.
+            var fieldBuilder = typeBuilder.DefineField($"_{name}", configTuple.Type, FieldAttributes.Private);
+
+            // Create the get method for the property that will return the field.
+            var getMethod = typeBuilder.DefineMethod($"get_{name}",
+                MethodAttributes.Public | MethodAttributes.HideBySig | MethodAttributes.SpecialName, configTuple.Type, null);
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs]: `parameterTypes: null`

File: tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs [hunk 23]
```diff
@@ -0,0 +1,454 @@
...
+                    configTuple.ExpectedValue == null ||
+                    (configTuple.ExpectedValue is string && string.IsNullOrEmpty((string) configTuple.ExpectedValue));
+                
+                var expectedIsInvalid = IsValueWrongType(configTuple.ExpectedValue, configTuple.Type);
+                var defaultIsInvalid = IsValueWrongType(configTuple.DefaultValue, configTuple.Type);
+
+                if ((configTuple.Required && expectedValueIsMissing) || 
+                    expectedIsInvalid || 
+                    defaultIsInvalid ||
+                    configTuple.ConfigKey == string.Empty)
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs]: Add a test for `null` `ConfigKey`.

File: tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs [hunk 24]
```diff
@@ -0,0 +1,454 @@
...
+        /// Used to test that the <see cref="ConfigurationFactory"/> fails correctly when given incorrect types.
+        /// </summary>
+        public class NoConversionFromThisClass
+        {
+        }
+
+        /// <summary>
+        /// Test data for the <see cref="ConfigurationFactory"/> tests.
+        /// </summary>
+        public static IEnumerable<object[]> ConfigurationFactoryTestData => new[]
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs]: Do you only support configuration values of primitive types?  What about nullables?  If yes, add two test cases:

1.  successful nullable conversion
2.  unsuccessful nullable conversion

File: tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs [hunk 25]
```diff
@@ -0,0 +1,454 @@
...
+        {
+            new object[]
+            {
+                // Succeeds
+                // Tests base case
+                new Dictionary<string, ConfigurationTuple>
+                {
+                    {
+                        "stringProperty",
+                        new ConfigurationTuple(typeof(string), false, "i am a string")
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs]: Named parameters would aid readability here and below.

File: tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+            Assert.Equal(secondSecret, value1);
+            Assert.Equal(value1, value2);
+        }
+
+        [Theory]
+        [InlineData(typeof(string))]
+        [InlineData(typeof(int))]
+        [InlineData(typeof(bool))]
+        [InlineData(typeof(DateTime))]
+        public async void HandlesKeyNotFound(Type type)
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs]: Prefer `Throws` instead of `Handles` in this case as it is more specific about the expected behavior.

Also `ThrowsForKeyNotFound` describes the scenario better.

You already do this [here](https://github.com/NuGet/ServerCommon/pull/12/files#diff-0beafe4a015085507da5f16982096ceeR140).

File: tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+            // GetOrDefault with default specified
+            Assert.Equal(memberOfType, await (dynamic)getOrDefaultMethod.Invoke(dummy, notFoundKeyDefaultSpecifiedArgs));
+        }
+
+        [Theory]
+        [InlineData(typeof(string))]
+        [InlineData(typeof(int))]
+        [InlineData(typeof(bool))]
+        [InlineData(typeof(DateTime))]
+        public async void HandlesNullOrEmptyArgument(Type type)
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs]: `ThrowsForNullOrEmptyArgument`

File: tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs [hunk 28]
```diff
@@ -0,0 +1,184 @@
...
+            // GetOrDefault with default specified
+            Assert.Equal(memberOfType, await (dynamic)getOrDefaultMethod.Invoke(configProvider, emptyKeyDefaultSpecifiedArgs));
+        }
+        
+        private class NoConversionFromStringToThisClass
+        {
+        }
+
+        [Fact]
+        public async void ThrowsNotFoundException()
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs]: `ThrowsForNonExistentSecretName`

File: tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs (hunk 23 above)
Comment (scottbommarito) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs]: I need to add more comments to the `ConfigurationTuple` class.

`ConfigKey` refers to the value of the `ConfigurationKeyAttribute` applied to the object, not the name of the property that is used. If `ConfigKey` is null, no `ConfigurationKeyAttribute` is applied and it uses the name of the property to access configuration, so `ConfigKey` being empty is the fail case. I can add a test for null `ConfigKey` in a separate test case.

File: tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs (hunk 24 above)
Comment (scottbommarito) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs]: Any type should work, I haven't done a nullable but I'll add one.

File: tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs (hunk 28 above)
Comment (scottbommarito) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs]: In retrospect, this test case doesn't even check `NotFound`, it checks `NotSupported`.

`ThrowsForNotSupportedConversion`

File: tests/NuGet.Services.Configuration.Tests/TestConfigurationProvider.cs [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+            string value;
+
+            if (!_configuration.TryGetValue(key, out value))
+            {
+                throw new KeyNotFoundException();
+            }
+
+            if (string.IsNullOrEmpty(value))
+            {
+                throw new ArgumentNullException();
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/TestConfigurationProvider.cs]: Yes, read [this](https://msdn.microsoft.com/en-us/library/system.argumentnullexception(v=vs.110).aspx#Remarks).

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs (hunk 15 above)
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: It's intended to use `configKey` so that one can easily find the the attribute that is malformed. In retrospect, I think all of the `ArgumentException`s need to use `configKey`.

File: src/NuGet.Services.Configuration/ConfigurationProvider.cs (hunk 17 above)
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/ConfigurationProvider.cs]: This method is not meant to throw so I will keep it as is. The call to `GetOrThrowAsync` will produce the `ArgumentNullException` indicating that the key is missing.

File: tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs (hunk 26 above)
Comment (scottbommarito) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs]: It doesn't always throw. `GetOrDefaultAsync` returns a default value instead of throwing. I named it `HandlesKeyNotFound` for this purpose. Is there a better name for that?

File: tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs (hunk 19 above)
Comment (scottbommarito) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs]: I remember I explicitly used `TypeDescriptor.GetConverter(type).ConvertFrom(value)` because I wanted to test when the types are convertible but the actual value is not.

E.G `"10"` can be converted to an int but `"this is a string"` cannot.

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs (hunk 3 above)
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: Added check below.

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs (hunk 14 above)
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: Added checks below.

File: src/NuGet.Services.Configuration/ConfigurationKeyPrefixAttribute.cs (hunk 8 above)
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/ConfigurationKeyPrefixAttribute.cs]: Added check below

File: src/NuGet.Services.Configuration/ConfigurationProvider.cs (hunk 16 above)
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/ConfigurationProvider.cs]: Added check below

File: tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs (hunk 24 above)
Comment (scottbommarito) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationFactoryFacts.cs]: added nullable tests below

File: tests/NuGet.Services.Configuration.Tests/TestConfigurationProvider.cs (hunk 29 above)
Comment (scottbommarito) [inline tests/NuGet.Services.Configuration.Tests/TestConfigurationProvider.cs]: I added `ConfigurationNullOrEmptyException`. There didn't seem to be a relevant pre-existing exception.

**Comment by scottbommarito** [comment]
@dtivel Next iteration is ready!

File: src/NuGet.Services.Configuration/ConfigurationKeyAttribute.cs [hunk 30]
```diff
@@ -13,6 +13,11 @@ public class ConfigurationKeyAttribute : Attribute
     {
         public ConfigurationKeyAttribute(string key)
         {
+            if (string.IsNullOrEmpty(key))
+            {
+                throw new ArgumentNullException(nameof(key));
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationKeyAttribute.cs]: Strictly speaking, you should throw an `ArgumentException` here with a message "Argument cannot be null or empty."  An `ArgumentException` handles the generic case, while `ArgumentNullException` handles a very narrow case.  Throwing an `ArgumentNullException` when key is a zero-length string is misleading.

File: src/NuGet.Services.Configuration/ConfigurationKeyPrefixAttribute.cs [hunk 31]
```diff
@@ -13,6 +13,11 @@ public class ConfigurationKeyPrefixAttribute : Attribute
     {
         public ConfigurationKeyPrefixAttribute(string prefix)
         {
+            if (string.IsNullOrEmpty(prefix))
+            {
+                throw new ArgumentNullException(nameof(prefix));
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationKeyPrefixAttribute.cs]: See previous comment.

File: src/NuGet.Services.Configuration/ConfigurationNullOrEmptyException.cs [hunk 32]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationNullOrEmptyException.cs]: Add copyright header.

File: src/NuGet.Services.Configuration/ConfigurationNullOrEmptyException.cs [hunk 33]
```diff
@@ -0,0 +1,27 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace NuGet.Services.Configuration
+{
+    public class ConfigurationNullOrEmptyException : Exception
```
Comment (dtivel) [inline src/NuGet.Services.Configuration/ConfigurationNullOrEmptyException.cs]: Add XML doc comments.

File: tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs [hunk 34]
```diff
@@ -80,64 +138,77 @@ public class ConfigurationProviderFacts
...
         }
-        
+
         private class NoConversionFromStringToThisClass
         {
         }
 
         [Fact]
-        public async void ThrowsNotFoundException()
+        public async void ThrowsForNotSupportedConversion()
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationProviderFacts.cs]: NotSupported => Unsupported

File: tests/NuGet.Services.Configuration.Tests/DictionaryConfigurationProvider.cs [hunk 35]
```diff
@@ -24,14 +24,19 @@ protected override Task<string> Get(string key)
         {
             string value;
 
+            if (string.IsNullOrEmpty(key))
+            {
+                throw new ArgumentNullException(key);
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/DictionaryConfigurationProvider.cs]: See previous comment about throwing the right exception type.

File: tests/NuGet.Services.Configuration.Tests/ConfigurationAttributeFacts.cs [hunk 36]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/ConfigurationAttributeFacts.cs]: Add copyright header.

File: tests/NuGet.Services.Configuration.Tests/DummyConfigurationProvider.cs [hunk 37]
```diff
@@ -0,0 +1,23 @@
...
+    /// <summary>
+    /// Returns "test value" for all configuration keys unless the given key is null or empty.
+    /// </summary>
+    public class DummyConfigurationProvider : ConfigurationProvider
+    {
+        protected override Task<string> Get(string key)
+        {
+            if (string.IsNullOrEmpty(key))
+            {
+                throw new ArgumentNullException(key);
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/DummyConfigurationProvider.cs]: See previous comment about exception type.

File: tests/NuGet.Services.Configuration.Tests/DummyConfigurationProvider.cs [hunk 38]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
```
Comment (dtivel) [inline tests/NuGet.Services.Configuration.Tests/DummyConfigurationProvider.cs]: Add copyright header.

File: src/NuGet.Services.Configuration/ConfigurationNullOrEmptyException.cs (hunk 33 above)
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/ConfigurationNullOrEmptyException.cs]: It is now thrown in `ConfigurationProvider`.

**Comment by scottbommarito** [comment]
@dtivel Second iteration done!

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs [hunk 39]
```diff
@@ -0,0 +1,135 @@
...
+                    throw new ArgumentException(
+                        $"Configuration key for {configKey} specified by {nameof(ConfigurationKeyAttribute)} is null or empty!",
+                        nameof(configNameProperty.Key));
+                }
+
+                configKey = configNameProperty.Key;
+            }
+
+            // Add the prefix specified by the ConfigurationKeyPrefixAttribute to the configuration key if it exists.
+            var configNamePrefixProperty =
```
Comment (skofman1) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: >configNamePrefixProperty [](start = 16, length = 24)

this can cause a problem in the gallery scenario, since all configuration items have the same prefix, and we don't want to add an identical attribute on all properties. I would suggest either having this attribute on the configuration class, or make it configurable on the factory level/input parameter.

**Comment by skofman1** [comment]
:shipit:

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs [hunk 40]
```diff
@@ -0,0 +1,135 @@
...
+
+        /// <summary>
+        /// Injects a property of <param name="instance">instance</param> specified by <param name="property">a <see cref="PropertyDescriptor"/></param> with configuration.
+        /// </summary>
+        /// <typeparam name="T">Type of <param name="instance">the instance</param>.</typeparam>
+        /// <typeparam name="TP">Type of <param name="property">the property</param>.</typeparam>
+        /// <param name="instance">Instance to inject configuration into a property of.</param>
+        /// <param name="property"><see cref="PropertyDescriptor"/> that describes the property to inject the configuration into.</param>
+        /// <returns>A task that completes when the property has been injected into.</returns>
+        public async Task InjectPropertyWithConfiguration<T, TP>(T instance, PropertyDescriptor property)
```
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: It's public because I use reflection to access it, and reflection can only access public methods.

File: src/NuGet.Services.Configuration/ConfigurationFactory.cs (hunk 39 above)
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/ConfigurationFactory.cs]: The `ConfigurationKeyPrefixAttribute` can now be applied to the entire `Configuration` class.
